Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp

```diff
@@ -6631,6 +6636,127 @@ void SpeculativeJIT::compileNewBoundFunction(Node* node)
+
+        JumpList notFastNamedCases;
+        // FIXME: Maybe we should have a better way to represent IndexedMode+OwnStructureMode?
+        bool indexedAndOwnStructureMode = m_graph.varArgChild(node, 1).node() == m_graph.varArgChild(node, 4).node();
+        JumpList& genericOrRecoverCase = indexedAndOwnStructureMode ? recoverGenericCase : notFastNamedCases;
+
+        // FIXME: We shouldn't generate this code if we know base is not an object.
+        notFastNamedCases.append(branchTest32(NonZero, modeGPR, TrustedImm32(JSPropertyNameEnumerator::IndexedMode | JSPropertyNameEnumerator::GenericMode)));
+        {
+            if (!m_state.forNode(baseEdge).isType(SpecCell))
+                genericOrRecoverCase.append(branchIfNotCell(baseRegs));
+
+            // Check the structure
+            // FIXME: If we know there's only one structure for base we can just embed it here.
+            load32(Address(baseRegs.payloadGPR(), JSCell::structureIDOffset()), scratchGPR);
+
+            genericOrRecoverCase.append(branch32(NotEqual, scratchGPR, Address(enumeratorGPR, JSPropertyNameEnumerator::cachedStructureIDOffset())));
+            emitNonNullDecodeZeroExtendedStructureID(scratchGPR, scratchGPR);
+            genericOrRecoverCase.append(branchTest32(NonZero, Address(scratchGPR, Structure::bitFieldOffset()), TrustedImm32(Structure::s_didWatchReplacementBits)));
+
+            // Compute the offset
+            // If index is less than the enumerator's cached inline storage, then it's an inline access
+            Jump outOfLineAccess = branch32(AboveOrEqual,
+                indexGPR, Address(enumeratorGPR, JSPropertyNameEnumerator::cachedInlineCapacityOffset()));
+
+            storeValue(valueRegs, BaseIndex(baseRegs.payloadGPR(), indexGPR, TimesEight, JSObject::offsetOfInlineStorage()));
+
+            doneCases.append(jump());
+
+            // Otherwise it's out of line
+            outOfLineAccess.link(this);
+            move(indexGPR, scratchGPR);
+            sub32(Address(enumeratorGPR, JSPropertyNameEnumerator::cachedInlineCapacityOffset()), scratchGPR);
+            neg32(scratchGPR);
+            signExtend32ToPtr(scratchGPR, scratchGPR);
+            loadPtr(Address(baseRegs.payloadGPR(), JSObject::butterflyOffset()), storageGPR);
+            constexpr intptr_t offsetOfFirstProperty = offsetInButterfly(firstOutOfLineOffset) * static_cast<intptr_t>(sizeof(EncodedJSValue));
+            storeValue(valueRegs, BaseIndex(storageGPR, scratchGPR, TimesEight, offsetOfFirstProperty));
+            doneCases.append(jump());
+        }
+
+        notFastNamedCases.link(this);
+        {
+            CodeOrigin codeOrigin = node->origin.semantic;
+            CallSiteIndex callSite = recordCallSiteAndGenerateExceptionHandlingOSRExitIfNeeded(codeOrigin, m_stream.size());
+            RegisterSetBuilder usedRegisters = this->usedRegisters();
+
+            auto [ stubInfo, stubInfoConstant ] = addStructureStubInfo();
```

Comment:
This is the same as the generic PutByVal case, right? Can we split this off into a helper?